File: 405.convert-a-number-to-hexadecimal.cpp
/*
 * @lc app=leetcode id=405 lang=cpp
 *
 * [405] Convert a Number to Hexadecimal
 */

#include <bits/stdc++.h>
#include <cassert>

using namespace std;

// @lc code=start
class Solution {
public:
    string toHex(int num) {
        if (num == 0) return "0";

        char hex[9] = {0};
        int32_t nums32 = num;
        uint32_t numu32 = (uint32_t)nums32;

        int i = 0;
        while(numu32) {
            char h = numu32 % 16;
            hex[7 - i++] = (h < 10) ? (h + '0') : (h - 10 + 'a');
            numu32 /= 16;
        }

        return std::string(hex + (8-i));
    }
};
// @lc code=end

int
main () {
  Solution s;
  assert(s.toHex(0) == "0");
  assert(s.toHex(26) == "1a");
  assert(s.toHex(-1) == "ffffffff");
  return 0;
}

File: 405.convert-a-number-to-hexadecimal.rs
/*
 * @lc app=leetcode id=405 lang=rust
 *
 * [405] Convert a Number to Hexadecimal
 */

// @lc code=start
impl Solution {
    pub fn to_hex(num: i32) -> String {
        let mut num = num as u32;
        if num == 0 {
            return "0".to_string();
        }
        let mut bytes = vec![];
        while num != 0 {
            let v = (num & 0xf) as u8;
            match v {
                0..=9 => bytes.push(b'0' + v),
                _ => bytes.push(b'a' + v - 10),
            }
            num >>= 4;
        }
        bytes.reverse();
        String::from_utf8(bytes).unwrap()
    }
}
// @lc code=end
